isign/bundle.py

```diff
@@ -198,71 +202,34 @@ class App(Bundle):
     signable_class = signable.Executable
 
     def __init__(self, path, native_platforms):
+        self.entitlements = None    # this is a bit ugly, but we have to communicate this down to Codesig
         super(App, self).__init__(path, native_platforms)
-        self.provision_path = join(self.path,
-                                   'embedded.mobileprovision')
-
-    def provision(self, provision_path):
-        log.debug("provisioning from {} to {}".format(provision_path, self.provision_path))
-        shutil.copyfile(provision_path, self.provision_path)
-
-    @staticmethod
-    def extract_entitlements(provision_path):
-        log.debug("EXTRACTING ENTITLEMENTS")
-        """ Given a path to a provisioning profile, return the entitlements
-            encoded therein """
-        cmd = [
-            'smime',
-            '-inform', 'der',
-            '-verify',  # verifies content, prints verification status to STDERR,
-            #  outputs content to STDOUT. In our case, will be an XML plist
-            '-noverify',  # accept self-signed certs. Not the opposite of -verify!
-            '-in', provision_path
-        ]
-        # this command always prints 'Verification successful' to stderr.
-        (profile_text, err) = openssl_command(cmd, data=None, expect_err=True)
-        if err and err.strip() != 'Verification successful':
-            log.error('Received unexpected error from openssl: {}'.format(err))
-        plist_dict = biplist.readPlistFromString(profile_text)
-        if 'Entitlements' not in plist_dict:
-            log.debug('failed to get entitlements in provisioning profile')
-            raise Exception('could not find Entitlements in {}'.format(provision_path))
-        return plist_dict['Entitlements']
-
-    def write_entitlements(self, entitlements):
-        """ Write entitlements to self.entitlements_path. This actually doesn't matter
-            to the app, it's just used later on by other parts of the signing process. """
-        entitlements_path = join(self.path,
-                                 'Entitlements.plist')
-        biplist.writePlist(entitlements, entitlements_path, binary=False)
-        log.debug("wrote Entitlements to {0}".format(entitlements_path))
-        self.entitlements_path = entitlements_path
-
-    def resign(self, deep, cms_signer, provisioning_profile, alternate_entitlements_path=None):
-        """ signs app in place """
 
-        # TODO all this mucking about with entitlements feels wrong. The entitlements_path is
-        # not actually functional, it's just a way of passing it to later stages of signing.
-        # Maybe we should determine entitlements data in isign/archive.py or even isign/isign.py,
-        # and then embed it into CmsSigner?
+    def provision(self, team_id, provisioner):
```

Comment:
I don't have a specific recommendation here, but when I see OO code that reads as a reflexive / redundant statement like "Bundle, provision yourself using this Provisioner", I always wonder if there's a clearer structure. I.e., why do Bundles know how to provision when there is a thing called a Provisioner? 

I dunno, maybe consider whether the method belongs on Provisioner, or whether Provisioner should be called something else if it's not the home for this method. I don't feel strongly about it.

Reply:
I agree with you and I think we're inching towards inverting control.

There are now so many options with signing that it might be better to have a SigningContext that operates on Bundles/Signables/etc, rather than handing Bundles a set of options and telling them to resign themselves. 

Condensing options into `CmsSigner` and now `Provisioner` has at least reduced the laundry list of things we are passing around.